scalameta/contrib/src/main/scala/scala/meta/contrib/implicits/Statements.scala

```diff
@@ -0,0 +1,105 @@
+package scala.meta.contrib.implicits
+
+import scala.collection.immutable.Seq
+import scala.language.higherKinds
+import scala.language.implicitConversions
+import scala.meta._
+import scala.meta.contrib.Equal
+
+/**
+  * Typeclass to deal with scalameta trees which have an expected set of statements.
+  *
+  * For trees that have multiple stat lists (such as Defn.Class) the main stats are used.
+  *
+  * This allows you to do things like add a statement to the
+  * body of a class without digging through the low level meta api
+  *
+  * TODO: Consider using a meta variant of simulacrum to
+  * generate the boilerplate once one is avaliable
+  */
+trait Statements[A] {
+  def getStats(a: A): Seq[Stat]
+  def replaceStats(a: A, newStats: Seq[Stat]): A
+}
+
+object Statements {
+
+  trait Ops[A] {
+    val self: A
+    val typeClassInstance: Statements[A]
+
+    def getStats: Seq[Stat] =
+      typeClassInstance.getStats(self)
+
+    def replaceStats(newStats: Seq[Stat]): A =
+      typeClassInstance.replaceStats(self, newStats)
+
+// DAVETODO: Get F param working (Aka Structurally/Syntactically)
+//    Is this going to be useful?
+//    def replaceStatWith[F[_ <: Tree]](stat: Stat, replacement: Stat)(
+//        implicit conv: Tree => F[Tree], eq: Equal[F[Tree]]): A = {
+//
+//      val newStats: Seq[Stat] = getStats.collect {
+//        case candidate if eq.equal(candidate, stat) => replacement
+//        case other => other
+//      }
+//
+//      replaceStats(newStats)
+//    }
+//    def removeStat[F[_]](stat: Stat)(implicit eq: Equal[F[A]]): A =
+//      removeStats[F](stat :: Nil)
+//
+//    def removeStats[F[_]](stats: Seq[Stat])(
+//      implicit conv: Tree => F[Tree], eq: Equal[F[Tree]]): A = {
+//
+//      val filtered: Seq[Stat] = getStats.filter(stat => stats.exists(eq.equals(_, stat)))
+//      replaceStats(filtered)
+//    }
+
+
+    def appendStat(stat: Stat): A =
```

Comment:
I agree on separating out Seq operations from getting a new tree with different stats.  Having multiple operations of type Seq[Stat] => A makes me want to separate out Seq[Stat] => Seq[Stat] and A.copy(stats = newStats)

I think many of these could be self.copy(stats = operationsFromSeqStatToSeqStat(originalStats)